[AL GHRAIRY ABDULELAH FAISAL S] iP

AAlghrairy opened the pull request:
# Wensleydale

> "A quote? What am I supposed to quote?" – Anonymous

Wensleydale is an interactive task-keeper that will ease your fears of forgetting. With it, you can:

- Interact in a GUI-based environment
- Key in important tasks and events in your day-to-day life
- Debate over intriguing philosophical questions and morally-ambiguous topics _(If all of those questions can be answered with an error message 😶 )_ 

To start, all you need to do is follow these easy steps:
1. Download the program from [here](https://github.com/AAlghrairy/ip)
2. Start the program
3. Begin adding your most important tasks and events ~~do not add sensitive information, program is not encrypted~~


Features include:
- [x] Documenting tasks
- [x] Search through list of tasks by keyword
- [x] Attributing dates and times to events
- [x] GUI-based environment 
- [ ] Manage overdue deadlines (coming soon)
- [ ] Sentience (not coming at all)   

Here's an example of the `main` method:

```java
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
```

Commit:
Add Gradle support

AAlghrairy committed:
Implement skeletal version

AAlghrairy committed:
Add list

AAlghrairy committed:
Add markable tasks

AAlghrairy committed:
Add ToDo, Events, Deadlines

AAlghrairy committed:
Add automated Text UI Testing

AAlghrairy committed:
Add Error Handling

AAlghrairy committed:
Update tasks to be deletable

AAlghrairy committed:
Use Enumeration

AAlghrairy committed:
Change Todo to ToDo

AAlghrairy committed:
Add saving feature

AAlghrairy committed:
Add date/time processing function

AAlghrairy committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

AAlghrairy committed:
Fix saving feature bug

AAlghrairy committed:
Refractor code to use more OOP

AAlghrairy committed:
Organize code into Packages

AAlghrairy committed:
Add JUnit tests

AAlghrairy committed:
Add Javadocs

AAlghrairy committed:
Reformat code to be compliant with the Coding Standard

AAlghrairy committed:
Add find method to search for keyword among the list of tasks

AAlghrairy committed:
Merge branch 'branch-A-JavaDoc'

AAlghrairy committed:
Merge branch 'branch-A-CodingStandard'

AAlghrairy committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

AAlghrairy committed:
Modify test-ui-test files

AAlghrairy committed:
Modify JavaDoc

AAlghrairy committed:
Fix Coding Standard violations

AAlghrairy committed:
Merge branch 'add-gradle-support' into branch-A-CodingStandard

AAlghrairy committed:
Add Gradle Support

AAlghrairy committed:
Merge branch 'branch-A-CodingStandard' into branch-A-Gradle

# Conflicts:
#	src/main/java/duke/Parser.java

AAlghrairy committed:
Fix Coding Standard violations

AAlghrairy committed:
Create GUI using JavaFX

AAlghrairy committed:
Add title to MainWindow

AAlghrairy committed:
Remove commented-out code

AAlghrairy committed:
Modify Duke.java constructor

AAlghrairy committed:
Modify code to coincide more with SLAP coding principle

AAlghrairy committed:
Remove references to "duke." package name in Strings

AAlghrairy committed:
Modify MainWindow to disable user input upon exiting the application

Spyobird commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        return null; //should not reach here
+    }
+
+    private void throwWrongToDoFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the description of a todo cannot be empty.");
+    }
+    private void throwWrongDeadlineFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the Deadline format is incorrect."
+            + " The format should be:\n\t[Task] [Description] /by yyyy-mm-dd/HH:mm (leave \"/HH:mm\""
+            + " empty if no time in current task)");
+    }
+
+    private void throwWrongEventFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the Event format is incorrect."
+            + " The format should be:\n\t[Task] [Description] /at yyyy-mm-dd/HH:mm/HH:mm");
+    }
+    private void throwEmptyDescriptionException() throws DukeException {
+        throw new DukeException("Pardon me, but the description of a deadline/event cannot be empty.");
+    }
+
+    private void throwEmptyDateException() throws DukeException {
+        throw new DukeException("Pardon me, but the date/time cannot be empty.");
+    }
```
Any reason why you decided to write a method for different messages for the same DukeException? Perhaps you could consider saving the error messages in to strings and calling `DukeException(ERROR_STRING)`. Alternatively, you could use subclasses of DukeException to handle different errors.

AAlghrairy replied:
That's a good point. I'll consider creating subclasses for this purpose.

Spyobird commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+                throw new DukeException("Pardon me, but the body of the find command should not be empty");
+            }
+
+            String foundTasks = tasks.find(keyword);
+
+            if (foundTasks.equals("")) {
+                returnMessage = "My apologies, but no tasks were found for the given keyword.";
+            } else {
+                returnMessage = "I've searched the archives and found these matching tasks in your list:\n"
+                    + tasks.find(keyword);
+            }
+            break;
+        case "todo":
+            currTask = parseMessageContents(message, TaskTypes.TODO);
+
+            tasks.addTask(currTask);
+            storage.modifyStorage(currTask, ConfirmCodes.ADDITION, tasks);
+
+            returnMessage = "Alright then! I've added the task to your list:" + "\n\t" + currTask
+                + "\n" + getTaskCountMessage(tasks);
+            break;
+        case "deadline":
+            currTask = parseMessageContents(message, TaskTypes.DEADLINE);
+            tasks.addTask(currTask);
+            storage.modifyStorage(currTask, ConfirmCodes.ADDITION, tasks);
+            returnMessage = "Alright then! I've added the task to your list:" + "\n\t" + currTask + "\n"
+                + getTaskCountMessage(tasks);
+            break;
+        case "event":
+            currTask = parseMessageContents(message, TaskTypes.EVENT);
+            tasks.addTask(currTask);
+            storage.modifyStorage(currTask, ConfirmCodes.ADDITION, tasks);
+            returnMessage = "Alright then! I've added the task to your list:" + "\n\t" + currTask + "\n"
+                + getTaskCountMessage(tasks);
+            break;
+        default:
+            throw new DukeException("Pardon me, but I did not understand what you said.");
+        }
+        return returnMessage;
+    }
```
If you have time, perhaps you can consider how to use abstraction and encapsulation to simplify repeated code. Long switch statements can be difficult to read at times.

AAlghrairy replied:
Thanks for the good responses. I've noticed you were not allocated to my code, so thank you very much for taking the time to review it 👍

Spyobird reviewed (commented):
Overall well written code. There is the occasional verbosity especially switch statements. Do consider how you can apply OOP principles such as abstraction and encapsulation which can help to cut down the verbosity and simplify the code.

9teMare commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,69 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate date;
+    protected LocalTime time;
```
I think it is better for these fields to be `private` and provide corresponding getter methods.

9teMare commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+     * @param input the user input
+     * @return the response to the user's input
+     */
+    public String getResponse(String input) {
```
Good naming convention here, good job 👍

aditi2313 replied:
Agreed!

9teMare commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate date;
+    protected LocalTime timeBeginning;
+    protected LocalTime timeEnd;
```
I think it is better for these fields to be `private` and provide corresponding getter methods. But very interesting idea to make a starting and ending time for `Event` 👍.

9teMare commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        }
+
+        switch (currMessage) {
+        case "help":
```
I think `help` is a great feature/command 👍.

9teMare commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+            timeEndString = messageArr[3];
+
+            date = parseDateFromString(dateString, TaskTypes.EVENT);
+            timeBegin = parseTimeFromString(timeBeginString, TaskTypes.EVENT);
+            timeEnd = parseTimeFromString(timeEndString, TaskTypes.EVENT);
```
Curious what will happen if the ending time is actually before the starting time? Is there any check for the validity of the duration?

AAlghrairy replied:
That's a really good point! Thank you for bringing that to my attention 👍

9teMare commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        return null; //should not reach here
+    }
+
+    private void throwWrongToDoFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the description of a todo cannot be empty.");
+    }
+    private void throwWrongDeadlineFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the Deadline format is incorrect."
+            + " The format should be:\n\t[Task] [Description] /by yyyy-mm-dd/HH:mm (leave \"/HH:mm\""
+            + " empty if no time in current task)");
+    }
+
+    private void throwWrongEventFormatException() throws DukeException {
+        throw new DukeException("Pardon me, but the Event format is incorrect."
+            + " The format should be:\n\t[Task] [Description] /at yyyy-mm-dd/HH:mm/HH:mm");
+    }
+    private void throwEmptyDescriptionException() throws DukeException {
+        throw new DukeException("Pardon me, but the description of a deadline/event cannot be empty.");
+    }
+
+    private void throwEmptyDateException() throws DukeException {
+        throw new DukeException("Pardon me, but the date/time cannot be empty.");
+    }
```
I think it would be better to change these methods into classes and package them under `DukeException`. Otherwise, you can simply `throw new DukeException()` with the corresponding error message.

9teMare commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @param list the list of tasks to be inputted
+     */
+    public TaskList(List<Task> list) {
```
I think `list` can be renamed to `tasksList` or simply `tasks`.

9teMare reviewed (commented):
LGTM 👍! 
- Great that you strictly followed the coding standard.
- I like your implementation of `enum`
- Just a minor concern that what will happen if for `Event` the ending time is earlier than starting time? I don't think I saw an implementation for this validity check.

AAlghrairy committed:
Update build.gradle to account for the change in Main class

AAlghrairy committed:
Add assertions to various classes

There are several important assumptions that should be documented that
are integral to program running without errors and/or bugs.

AAlghrairy committed:
Improve code quality

There are several areas of the code which consist of verbose statements,
as well as areas where the code was reused and redundant.

As a step towards better code quality, the redundant code was removed
and the code was refactored to coincide more with SLAP coding principles
for the sake of code readability.

AAlghrairy committed:
Merge pull request #1 from AAlghrairy/branch-A-Assertions

Add assertions to various classes

AAlghrairy committed:
Merge branch 'master' of https://github.com/AAlghrairy/ip

* 'master' of https://github.com/AAlghrairy/ip:
  Add assertions to various classes

AAlghrairy committed:
Merge branch 'master' into branch-A-CodeQuality

AAlghrairy committed:
Merge pull request #2 from AAlghrairy/branch-A-CodeQuality

Improve code quality

AAlghrairy committed:
Merge branch 'master' of https://github.com/AAlghrairy/ip

* 'master' of https://github.com/AAlghrairy/ip:
  Improve code quality

AAlghrairy committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/AAlghrairy/ip into branch-A-CodeQuality

* 'branch-A-CodeQuality' of https://github.com/AAlghrairy/ip:
  Add assertions to various classes

AAlghrairy committed:
Merge branch 'branch-A-CodeQuality'

AAlghrairy committed:
Add more assertions to Parser.java

AAlghrairy committed:
Add sorting ability

A command to sort the list of tasks.

There are several ways to sort the list of tasks:

*Chronologically
*Alphabetically (by description)
*Marked (whether they are done or not

AAlghrairy committed:
Modify GUI design to resize text-fields automatically

AAlghrairy committed:
Modify GUI design

eugenechiaay commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("Pardon me, but the provided index does not exist in your list.");
+        }
+    }
```
Nice naming here, an upgrade over the simple "get()"

aditi2313 replied:
Agreed 😁

eugenechiaay commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Returns the number of tasks currently in the list.
+     */
+    public int getTasksCount() {
```
Nice naming here too, an upgrade over the simple "size()"

eugenechiaay reviewed (commented):
Overall very well written code. Can't really find any coding standard violation tbh...

aditi2313 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,121 @@
+ */
+public class Event extends Task {
+
+    public static final String FORMAT = "[Task] [Description] /at yyyy-mm-dd/HH:mm/HH:mm";
```
Good job using all caps for this static variable!

aditi2313 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,149 @@
+     * @throws DukeException throws if text file does not exist
+     * @throws IOException   throws if text file cannot be created
+     */
+    public List<Task> load() throws DukeException, IOException {
```
Method name is a verb and complies with the Java naming convention! Good job keeping that in mind! 😀

aditi2313 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class Ui {
+
+    protected static final String HELP = "Here are the list of commands:\n"
+        + "|list [index]|: Show all the tasks currently in the list\n"
+        + "|mark [index]|: Mark a task as done\n"
+        + "|unmark [index]|: Mark a task as undone\n"
+        + "|delete [index]|: Delete a task\n"
+        + "|find [keyword]|: Show the list of tasks that match keyword\n"
+        + "|todo [description]|: Add a ToDo task. Represents a task without a date/time\n"
+        + "|deadline [Description] /by yyyy-mm-dd|: Add a Deadline Task. Represents a task that must be"
+        + "\n\tcompleted by certain date\n"
+        + "|deadline [Description] /by yyyy-mm-dd/HH:mm|: Add a Deadline task with a time condition.\n"
+        + "|event [Description] /at yyyy-mm-dd/HH:mm/HH:mm|: Add an Event task. Represents a task that"
+        + "\n\tstarts within a specific time range on a specific day "
+        + "\n\t(HH:mm/HH:mm represents Begin/End time)\n"
+        + "|sort [keyword]| sorts the list using the given keyword:\n"
+        + "\tchronologically: sorts tasks by date and time\n"
+        + "\talphabetically: sorts tasks alphabetically by description\n"
+        + "\tdone: sorts tasks by whether they are done or not\n";;
```
Good use of indentation here! :)

AAlghrairy replied:
Thank you very much for your helpful and encouraging review comments 👍

aditi2313 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,72 @@
+     * @param input   the user input
+     * @return the chatbot response
+     */
+    public String userInput(TaskList tasks, Storage storage, String input) {
```
Perhaps make this method name a verb to keep it consistent with Java's naming convention?

AAlghrairy replied:
Thanks for reminding me 👍

aditi2313 commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,67 @@
+
+public class ParserTest {
+    @Test
+    public void processMessage_wrongFormatMessage_exceptionThrown() throws IOException {
```
Maybe consider making these method names shorter? 😅

AAlghrairy replied:
To clarify, this method name is long as it's a test method, so I need to follow the naming conventions of test methods:

![image](https://user-images.githubusercontent.com/77106335/152783697-dcdeb6cc-bc12-402a-851c-503a171cd66a.png)

But I will try to modify it to make the name shorter and more explicit.

aditi2313 reviewed (commented):
Overall, pretty impressed with how well your code follows the Java coding standard and naming convention. Just remember to make some of your method names a little shorter and use verbs instead. But otherwise, good job! 👍🏻

AAlghrairy committed:
Modify the naming of some methods to better follow Naming Conventions

AAlghrairy committed:
Modify Parser.java to be shorter

AAlghrairy committed:
Modify Parser.java to fix bugs

AAlghrairy committed:
Improve format-checking of input in Parser.java

AAlghrairy committed:
Improve format-checking in Parser.java with regards to whitespace check

AAlghrairy committed:
Improve the GUI and modify code to better follow naming conventions

AAlghrairy committed:
Modify variable names to be more explicit

AAlghrairy committed:
Add a User-Guide and representative screenshot

AAlghrairy committed:
Modify the README.md and rename Ui.png

AAlghrairy committed:
Modify the README.md

AAlghrairy committed:
Modify Ui.png

AAlghrairy committed:
Update Ui.png

AAlghrairy committed:
Modify README.md

AAlghrairy committed:
Delete leftover To-Do comment

AAlghrairy committed:
Set theme jekyll-theme-slate

AAlghrairy committed:
Update README.md

AAlghrairy committed:
Merge branch 'master' of https://github.com/AAlghrairy/ip

AAlghrairy committed:
Update README.md

AAlghrairy committed:
Update README.md to include additional details

AAlghrairy committed:
Fix Java Coding Standard violations

AAlghrairy committed:
Fix more Java coding standard violations

AAlghrairy committed:
Revert "Fix more Java coding standard violations"

This reverts commit a16cc0948b580f413d81f8de626272c03583fa88.

AAlghrairy committed:
Fix Java coding standard violations

AAlghrairy committed:
Fix bug in sort chronologically function and shorten long methods
